### DISCUSSION TITLE
Modifying .reference elements in Patient record output

### DISCUSSION THREAD

**Discussion by ToInf1n1tyAndBeyond** [synthetichealth/synthea#1331] [Q&A] [answered]
We are attempting to upload a large number of Synthea patient data into a FHIR server; however, we have discovered that all reference url's in the output are invalid and as a result are rejected by the FHIR profile. Solution is to either remove all .reference elements from the .json files or somehow have the program generate valid reference url's. Worst-case scenario would be to modify the core code to suppress the .reference elements. Can someone advise if...

1. There is a simpler way to suppress the .reference elements in the .json output, or
2. There is some way to have the code generate valid reference url's

**Comment by jawalonoski** [comment]
We regularly test this output with the HAPI server, so it would be useful if you can go into the specifics of your situation.

Synthea has a number of ways you can alter the FHIR output, specifically (abridged `synthea.properties` file):

```properties
# exporters that use XML or JSON can enable or disable 'pretty printing'
exporter.pretty_print = true
exporter.fhir.transaction_bundle = true
# using bulk_data=true will ignore exporter.pretty_print
exporter.fhir.bulk_data = false
```

- When `exporter.fhir.bulk_data` is `true` then _most_ references appear as `{ResourceType}/{ResourceID}`.
  - In this case, each `Patient` and their associated Resources are contained in a `Bundle`
  - The `exporter.fhir.transaction_bundle` setting controls whether the `Bundle` is a `collection` or `transaction`.
    - If it is a `transaction` then the references of resources shared across `Patient` records (i.e., `Organization`, `Location`, `Practitioner`) are constructed as queries, e.g., `Organization?identifier=https://github.com/synthetichealth/synthea|{ResourceID}`. In this case, you need to load the `hospitalInformation*.json` and `practitionerInformation*.json` files _before_ any of the patient records.
    - If your system cannot handle loading the shared resources in advance, and querying for the resource in the transaction, then use `collection` -- just note that shared resources will be duplicated.
- When `exporter.fhir.bulk_data` is `false` then references appear as `urn:uuid:{ResourceID}`
  - In this case, each `Patient` and their associated Resources are written into separate ndjson files. 
  - Shared resources should be loaded in advance... for example, loading `Organization` before `Location` before `Practitioner` etc.

**Comment by ToInf1n1tyAndBeyond** [comment]
Attached is a .json with generated patient and encounter data. When trying to upload to my Smile FHIR instance all references generate errors...

- "HAPI-1091: Invalid match URL \"Practitioner?identifier=http://hl7.org/fhir/sid/us-npi|9999869198\" - No resources match this search"
- "HAPI-1091: Invalid match URL \"Organization?identifier=https://github.com/synthetichealth/synthea|17727201-cc60-3852-b377-9ae26b087fe0\" - No resources match this search"
- "HAPI-1091: Invalid match URL \"Practitioner?identifier=http://hl7.org/fhir/sid/us-npi|9999998393\" - No resources match this search"
- "HAPI-1091: Invalid match URL \"Organization?identifier=https://github.com/synthetichealth/synthea|c2b4bc00-156e-3478-a63f-4561dbab7e26\" - No resources match this search"

When .reference elements are removed the remaining data is uploaded successfully.

[Sharon747_Fahey393_ca617c5f-828d-9a38-7e5c-41cb824c5ea8.txt](https://github.com/synthetichealth/synthea/files/12043697/Sharon747_Fahey393_ca617c5f-828d-9a38-7e5c-41cb824c5ea8.txt)

**Reply by jawalonoski** [reply, marked as answer]
Did you load the `hospitalInformation*.json` and `practitionerInformation*.json` files first?

Those two files contain the shared dependencies (Organization, Location, Practitioner) that I described above.

**Reply by ToInf1n1tyAndBeyond** [reply]
I didn't! Thank you.
